smart-contracts/ethereum/contracts/MarketPlace.sol

```diff
@@ -1,43 +1,105 @@
- */
-contract Marketplace {
-  modifier onlyOperator() {
-    require(operators[msg.sender]);
+pragma solidity >=0.4.15;
+
+contract MarketPlace {
+
+  modifier onlyDelegates() {
+    require(delegates[msg.sender]);
     _;
   }
 
-  struct appVersion {
-    uint version;
+  struct AppVersion {
+    bytes32 version;
     string files;
-    string category;
-    string status;
+    mapping (address => int8) votes;
+    address[] voters; // we might exclude old voters after elections
     string checksum;
-    uint required_permissions;
+    uint permissions;
   }
+  
 
   struct App {
     uint id;
     address owner;
     string name;
-    appVersion[] versions;
+    string category;
+    mapping  (bytes32 => AppVersion)  versions;
+    bytes32[] versionList;
   }
   
   mapping (uint => App) apps;
-  mapping (address => bool) operators;
+  uint[] app_ids;
+  mapping (address => bool) delegates;
+
+  address republicContract;
+  
+  event AppSubmitted(uint id);
+  event VersionSubmitted(uint app_id, bytes32 version);
+  event AppVoting(uint app_id, bytes32 version, int8 vote);
+  
+  function MarketPlace() public {
+    delegates[msg.sender] = true; // only one delegate for now until republic contract is fully functional
+  }
+
+  function setRepublicContract (address _contract_address) public onlyDelegates returns(bool res) {
+    republicContract = _contract_address;
+
+    return true;
+  }
+  
+  function refreshDelegatesList() public pure returns(bool res) { // this should be called after delegates elections
+    return true;
+  }
   
-  function Marketplace() public {
-    
+
+  function submitAppForReview(string _name, bytes32 _version, string _category, string _files, string _checksum, uint _permissions) public returns(bool res) {
+    uint new_id = app_ids.length;
+    app_ids.push(new_id);
+    bytes32[] memory versionList;
+
+    apps[new_id] = App(new_id, msg.sender, _name, _category, versionList);
+    AppSubmitted(new_id);
+    submitVersionForReview(new_id, _version, _files, _checksum, _permissions);
+
+    return true;
   }
 
-  function submitAppForReview() public pure returns(bool res) { // TODO: string name, uint version, string files, string checksum
+  function submitVersionForReview(uint _id, bytes32 _version, string _files, string _checksum, uint _permissions) public returns(bool res) {
+    require(apps[_id].id == _id);
+    require(apps[_id].versions[_version].version != _version);
+    address[] memory voters;
+
+    apps[_id].versions[_version] = AppVersion(_version, _files, voters, _checksum, _permissions);
+    apps[_id].versionList.push(_version);
+    VersionSubmitted(_id, _version);
+
     return true;
   }
+  
+
 
-  function vote() onlyOperator public view returns(bool res) { // TODO: uint id, bool vote
+  function voteForApp(uint _id, bytes32 _version, int8 _vote) public onlyDelegates returns(bool res) {
+    require (apps[_id].id == _id && apps[_id].versions[_version].version == _version);
```

Comment:
Kind of a crazy check, what about a comment "Check if the app and requested version exists"

Reply:
sure! will add some comments
(side joke: implementation is the documentation 😜 )